sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,312 @@
+        TokenRange tokenRange =
+            new TokenRange(
+                row.getLong("partitions_count"),
+                row.getLong("mean_partition_size"),
+                row.getString("range_start"),
+                row.getString("range_end"));
+        tokenRanges.add(tokenRange);
+      }
+      // The table may not contain the estimates yet
+      // or have partitions_count and mean_partition_size fields = 0
+      // if the data was just inserted and the amount of data in the table was small.
+      // This is very common situation during tests,
+      // when we insert a few rows and immediately query them.
+      // However, for tiny data sets the lack of size estimates is not a problem at all,
+      // because we don't want to split tiny data anyways.
+      // Therefore, we're not issuing a warning if the result set was empty
+      // or mean_partition_size and partitions_count = 0.
+      return tokenRanges;
+    } catch (InvalidQueryException e) {
+      LOG.warn(
+          "Failed to fetch size estimates for $keyspaceName.$tableName "
+              + "from system.size_estimates table. "
+              + "The number of created bundles may be inaccurate. "
+              + "Please make sure you use Cassandra 2.1.5 or newer.");
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      tokenRanges.add(new TokenRange(0L, 0L, "0", "0"));
+      return tokenRanges;
+    }
+  }
+
+  /**
+   * values that we get from system.size_estimates are for one node. We need to extrapolate to the
+   * whole cluster. This method estimates the percentage, the node represents in the cluster.
+   */
+  private static float getRingFraction(Cluster cluster, List<TokenRange> tokenRanges) {
+    Partitioner partitioner = new Partitioner(cluster);
+    long addressedTokens = 0;
+    for (TokenRange tokenRange : tokenRanges) {
+      addressedTokens =
+          addressedTokens + distance(cluster, tokenRange.rangeStart, tokenRange.rangeEnd);
+    }
+    // it is < 1 because it is a percentage
+    return Float.valueOf(addressedTokens / (partitioner.getTotalTokens()));
+  }
+
+  /**
+   * Measure distance between two tokens.
+   */
+  private static long distance(Cluster cluster, long token1, long token2) {
+    Partitioner partitioner = new Partitioner(cluster);
+    if (token2 > token1) {
+      return token2 - token1;
+    } else {
+      return (token2 - token1 + partitioner.getTotalTokens());
+    }
+  }
+
+  static class Partitioner {
+    long minToken;
+    long maxToken;
+
+    Partitioner(Cluster cluster) {
+      String partitionerName = cluster.getMetadata().getPartitioner();
+      switch (partitionerName) {
+        case "org.apache.cassandra.dht.Murmur3Partitioner":
+          minToken = Long.MIN_VALUE;
+          maxToken = Long.MAX_VALUE;
+          break;
+        case "org.apache.cassandra.dht.RandomPartitioner":
+          minToken = -1;
+          maxToken = (long) Math.pow(2, 127);
```

Comment:
I'm not sure where my comments about this disappeared - can't find them on the PR anymore...

This code path is definitely wrong and will produce data loss. Math.pow(2, 127) is not representable as a long, since long is only 64 bits and not 128 - so you'll just get Long.MAX_VALUE here, according to Java's casting rules.

I also don't understand why minToken should be -1 here.

Why not always use Long.MIN_VALUE and Long.MAX_VALUE, and get rid of the Partitioner class altogether?